### PR TITLE
Add proper error handling for memory allocation failures

### PR DESCRIPTION
**Pull request by shmolyneaux**
This PR converts the memory allocator and related functions from panicking on allocation failure to returning `Result` types, enabling graceful error handling when the heap is exhausted.

## Summary
Previously, memory allocation failures would cause a panic. This change propagates allocation errors throughout the codebase as `Result<T, String>` return types, allowing the interpreter to report out-of-memory conditions as clean errors rather than crashing.

## Key Changes

**Memory Management (`mem.rs`)**
- `alloc_no_debug()`, `alloc_debug()`, `alloc_and_set()`, `alloc_str_raw()` now return `Result<u24, String>` instead of panicking
- All allocation methods that depend on these (e.g., `alloc_str()`, `alloc_dict()`, `alloc_list()`, `alloc_tuple()`, `alloc_fn()`) updated to return `Result`
- Out-of-memory errors now produce user-friendly messages instead of panics

**Runtime (`runtime.rs`)**
- `Environment::realloc()` returns `Result<u24, String>`
- `Environment::insert_new()` and `push_scope()` return `Result<(), String>`
- Builtin registration during interpreter setup uses `.expect()` since failures there are unrecoverable
- Struct instantiation and other allocation-dependent operations now propagate errors

**Shimlibs (`shimlibs.rs`)**
- All iterator and native type methods that allocate now propagate `Result` types
- `ShimDict::expand_capacity()`, `realloc_entries()`, `clear_and_alloc_indices()`, and `shrink_to_fit()` return `Result<(), String>`
- `ShimList::push()` returns `Result<(), String>`
- All `?` operators added to propagate allocation failures up the call stack

**Parser & Compiler**
- Integer literals now carried as `i64` to support `i32::MIN` (-2147483648) via unary minus folding
- Out-of-range integer literals (e.g., bare `2147483649`) rejected at parse time with clear error messages
- Bare block expressions (used as statements) no longer require trailing semicolons

**Language Features**
- Added `sep` and `end` keyword arguments to `print()` function
- Comparison operators (`<`, `>`, `<=`, `>=`) and `in` operator now support struct overloads that return truthy/falsy values
- Improved documentation for string limits (65,535 bytes max), modulo behavior with negative divisors, and non-finite float handling

## Notable Implementation Details

- The `alloc!` macro throughout the codebase now requires `?` to propagate errors
- Initialization code that runs during interpreter setup uses `.expect()` since OOM at startup is unrecoverable
- Integer literal range validation happens at both lex time (for obviously out-of-range values) and compile time (for edge cases)
- Test cases added for out-of-memory conditions, string length limits, and new language features

https://claude.ai/code/session_01HTubigc2D7AzR2jPKu5nTN

### CODE REVIEW

File: shimlang/test_scripts/errors/string_too_long.stderr [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ 2 | let i = 0;
+ 3 | while i < 25 {
+ 4 |     s = s + s;
+               ^
+ 5 |     i += 1;
+ 6 | }
```
Comment (shmolyneaux) [inline shimlang/test_scripts/errors/string_too_long.stderr:6]: Use `for _ in 0..25 {...}`

File: shimlang/test_scripts/errors/string_too_long.shm [hunk 2]
```diff
@@ -0,0 +1,7 @@
+let s = "x";
+let i = 0;
+while i < 25 {
+    s = s + s;
+    i += 1;
+}
```
Comment (shmolyneaux) [inline shimlang/test_scripts/errors/string_too_long.shm:6]: Use for _ in 0..25 {...}
